### PR TITLE
perf(cli): fast path common argument patterns

### PR DESCRIPTION
**Pull request by mmastrac**
```

15:20 $ hyperfine -S none --warmup 25 '/tmp/deno run /tmp/empty.js' 'target/release/deno run /tmp/empty.js'
Benchmark 1: /tmp/deno run /tmp/empty.js
  Time (mean ± σ):      17.6 ms ±   5.3 ms    [User: 11.4 ms, System: 4.2 ms]
  Range (min … max):    15.1 ms …  85.0 ms    169 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 2: target/release/deno run /tmp/empty.js
  Time (mean ± σ):      16.4 ms ±   0.7 ms    [User: 10.8 ms, System: 4.0 ms]
  Range (min … max):    14.6 ms …  20.1 ms    193 runs
 
Summary
  'target/release/deno run /tmp/empty.js' ran
    1.07 ± 0.33 times faster than '/tmp/deno run /tmp/empty.js'
```

### CODE REVIEW

File: cli/args/flags.rs [hunk 1]
```diff
@@ -895,6 +895,49 @@ To evaluate code in the shell:
...
+      return Ok(Flags {
+        subcommand: DenoSubcommand::Run(RunFlags {
+          script,
+          ..Default::default()
+        }),
+        ..Default::default()
+      });
+    }
+    if args.len() == 4
+      && (args[2] == "-A" || args[2] == "--allow")
```
Comment (lucacasonato) [inline cli/args/flags.rs:926]:
```suggestion
      && (args[2] == "-A" || args[2] == "--allow-all")
```

Careful!

File: cli/args/flags.rs [hunk 2]
```diff
@@ -895,6 +895,49 @@ To evaluate code in the shell:
 
 /// Main entry point for parsing deno's command line flags.
 pub fn flags_from_vec(args: Vec<String>) -> clap::error::Result<Flags> {
+  // Fast path: "deno"
+  if args.len() < 2 {
+    let mut flags = Flags {
```
Comment (lucacasonato) [inline cli/args/flags.rs:900]: Does flags also parse env vars? Need to make sure the clap flag parsing is actually pure

File: cli/args/flags.rs (hunk 1 above)
Comment (mmastrac) [reply cli/args/flags.rs:926]: Ack. good catch here

File: cli/args/flags.rs (hunk 2 above)
Comment (mmastrac) [reply cli/args/flags.rs:900]: I see three env vars in the flag parsing -- those will need to be dealt with.

File: cli/args/flags.rs [hunk 3]
```diff
@@ -971,6 +1031,60 @@ pub fn flags_from_vec(args: Vec<String>) -> clap::error::Result<Flags> {
   Ok(flags)
 }
 
+/// The fast flag parser for the most common options.
+fn fast_flags_from_vec(args: &[String], env_vars: &EnvVars) -> Option<Flags> {
+  // We don't parse environment variables in this path, just abort if we see any
+  // that impact parsing.
+  if !env_vars.is_empty() {
```
Comment (mmastrac) [inline cli/args/flags.rs:1038]: @lucacasonato Added a bailout if we see env vars -- also these env vars are parsed in a single, consistent place .

File: cli/args/flags.rs [hunk 4]
```diff
@@ -4298,6 +4401,41 @@ mod tests {
...
 
+  #[test]
+  fn fast_path() {
+    for args in [
+      svec!["deno"],
+      svec!["deno", "run", "script.js"],
+      svec!["deno", "run", "--allow-all", "script.ts"],
+      svec!["deno", "run", "-A", "script.ts"],
+    ] {
+      assert_eq!(
```
Comment (mmastrac) [inline cli/args/flags.rs:4412]: Explicit test to ensure that we parse all fast-path commands the same as we'd parse with clap.

File: cli/args/flags.rs [hunk 5]
```diff
@@ -893,8 +905,50 @@ To evaluate code in the shell:
...
+      }
+
+      pub fn is_empty(&self) -> bool {
+        $(self.$arg.is_none()) && *
+      }
+    }
+  }
+}
+
+env_args!(
```
Comment (mmastrac) [inline cli/args/flags.rs:930]: This is not comprehensive by any means, but we can grow this over time.

File: cli/args/flags.rs [hunk 6]
```diff
@@ -1025,17 +1158,15 @@ fn clap_root() -> Command {
...
+        .help("Enable unstable bare node builtins feature (or set DENO_UNSTABLE_BARE_NODE_BUILTINS=true).")
         .value_parser(FalseyValueParser::new())
         .action(ArgAction::SetTrue)
         .global(true),
     )
     .arg(
       Arg::new("unstable-byonm")
         .long("unstable-byonm")
-        .help("Enable unstable 'bring your own node_modules' feature")
-        .env("DENO_UNSTABLE_BYONM")
```
Comment (lucacasonato) [inline cli/args/flags.rs:1038]: Not including this here means it doesn't show in --help anymore

Comment (bartlomieju) [reply cli/args/flags.rs:1038]: They need to manually added to `ENV_VARIABLES_HELP`

**Pull request closed without merging**